src/main/host/descriptor/pipe.rs

```diff
@@ -301,26 +310,43 @@ impl PipeFile {
         self.state
     }
 
-    fn filter_state(&self, mut state: FileState) -> FileState {
-        // if not open for reading, remove the readable flag
-        if !self.mode.contains(FileMode::READ) {
-            state.remove(FileState::READABLE);
+    /// Align the pipe's state to the buffer state. For example if the buffer is both `READABLE` and
+    /// `WRITABLE`, and the pipe is only open in `READ` mode, the pipe's `READABLE` state will be
+    /// set and the `WRITABLE` state will be unchanged.
+    fn align_state_to_buffer(&mut self, buffer_state: BufferState, event_queue: &mut EventQueue) {
+        let mut mask = FileState::empty();
+        let mut file_state = FileState::empty();
+
+        // if the pipe is already closed, do nothing
+        if self.state.contains(FileState::CLOSED) {
+            return;
         }
 
-        // if not open for writing, remove the writable flag
-        if !self.mode.contains(FileMode::WRITE) {
-            state.remove(FileState::WRITABLE);
+        // only update the readable state if the file is open for reading
+        if self.mode.contains(FileMode::READ) {
+            mask.insert(FileState::READABLE);
+            // file is readable if the buffer is readable or there are no writers
+            if buffer_state.intersects(BufferState::READABLE | BufferState::NO_WRITERS) {
+                file_state.insert(FileState::READABLE);
+            }
         }
 
-        state
+        // only update the writable state if the file is open for writing
+        if self.mode.contains(FileMode::WRITE) {
+            mask.insert(FileState::WRITABLE);
+            // file is writable if the buffer is writable or there are no readers
+            if buffer_state.intersects(BufferState::WRITABLE | BufferState::NO_READERS) {
+                file_state.insert(FileState::WRITABLE);
```

Comment:
What happens when a write operation is performed on a pipe with no readers? Is the data discarded? Do we have a test case for this?

Reply:
The write will return `EPIPE` (and I think also signal `SIGPIPE`, but we don't handle `SIGPIPE` yet in shadow and I think CMake ignores `SIGPIPE` in all tests like the non-shadow/linux tests). This should be covered by the two tests `test_write_after_read_close_with_full_buffer` and `test_write_after_read_close_with_nonfull_buffer`.